### PR TITLE
DB setup

### PR DESCRIPTION
**Pull request by jah3**

### CODE REVIEW

**Comment by taragrammatonn** [comment]
resolve conflicts before rebase and merge this request

**Comment by taragrammatonn** [comment]
based on commit description, you should have only DB setup, not controller and services layer, please add separate commits for it

File: src/test/java/md/maib/CustomerControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Test
+    void shouldGetAllCustomers() {
+        List<Customer> customers = List.of(
+                new Customer(null, "John", "john@mail.com", "123"),
+                new Customer(null, "Dennis", "dennis@mail.com", "456")
+        );
+        customerRepository.saveAll(customers);
+
+        given()
```
Comment (taragrammatonn) [inline src/test/java/md/maib/CustomerControllerTest.java:65]: please use tabs for easier reading

File: src/test/java/md/maib/CustomerControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                .contentType(ContentType.JSON)
+                .when()
+                .get("/api/customers")
+                .then()
+                .statusCode(200)
+                .body(".", hasSize(2));
```
Comment (taragrammatonn) [inline src/test/java/md/maib/CustomerControllerTest.java:71]: the test is not entirely accurate, here is another approach

```java
//@formatter:off
        var customerList = 
                given()
                    .contentType(ContentType.JSON)
                .when()
                    .get("/api/customers")
                .then()
                    .statusCode(200)
                    .extract()
                    .body()
                    .jsonPath()
                    .getList(".", Customer.class);

        assertThat(customers)
                .usingRecursiveComparison()
                .ignoringFields("id")
                .isEqualTo(customerList);
//@formatter:on
```

File: src/test/java/md/maib/CustomerControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    @BeforeEach
+    void setUp() {
+        RestAssured.baseURI = "http://localhost:" + port;
+        customerRepository.deleteAll();
+    }
+
+    @Test
+    void shouldGetAllCustomers() {
+        List<Customer> customers = List.of(
```
Comment (taragrammatonn) [inline src/test/java/md/maib/CustomerControllerTest.java:59]: use var

File: src/main/java/md/maib/entity/Customer.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    @Column(name = "username", nullable = false, unique = true)
+    private String name;
+
+    @Column(name = "pan", nullable = false, unique = true)
+    private String email;
+
+    @Column(name = "cvv", nullable = false)
+    private String cvv;
+
+    public Customer() {
```
Comment (taragrammatonn) [inline src/main/java/md/maib/entity/Customer.java:28]: should be on top

**Review by taragrammatonn — CHANGES_REQUESTED** [review]

File: src/main/resources/schema.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE users (
+                       user_id SERIAL PRIMARY KEY,
+                       username VARCHAR(50)  NOT NULL,
+                       pan VARCHAR(16) UNIQUE NOT NULL,
+                       cvv VARCHAR(3) NOT NULL
+);
+CREATE TABLE transactions (
```
Comment (taragrammatonn) [inline src/main/resources/schema.sql:7]: transactions is not represened in ORM

File: src/main/resources/schema.sql [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+                              transaction_id SERIAL PRIMARY KEY,
+                              user_id INT REFERENCES users(user_id),
```
Comment (taragrammatonn) [inline src/main/resources/schema.sql:9]: @Sp1ithy is not better to use BIGINT here?

**Review by taragrammatonn — CHANGES_REQUESTED** [review]

File: src/main/java/md/maib/entity/Customer.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+
+@Entity
+@Table(name = "users")
+public class Customer {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
[the remark this answers is not in the document]
Comment (jah3) [reply src/main/java/md/maib/entity/Customer.java:15]: it allows the database to automatically generate unique identifiers for each record

File: src/main/java/md/maib/controller/CustomerController.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package md.maib.controller;
+
+import java.util.List;
+
+import md.maib.entity.Customer;
+import md.maib.repository.CustomerRepository;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (Sp1ithy) [inline src/main/java/md/maib/controller/CustomerController.java:10]: How is it common with **DB Setup** ?

File: src/test/java/md/maib/CustomerControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+import org.testcontainers.containers.PostgreSQLContainer;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.hasSize;
+
+import java.util.List;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CustomerControllerTest {
```
Comment (Sp1ithy) [inline src/test/java/md/maib/CustomerControllerTest.java:25]: Same, how is it common with **DB setup** ?

**Review by Sp1ithy — CHANGES_REQUESTED** [review]

File: src/main/java/md/maib/controller/CustomerController.java (hunk 8 above)
Comment (jah3) [reply src/main/java/md/maib/controller/CustomerController.java:10]: i checked if the testcontainer works

Comment (Sp1ithy) [reply src/main/java/md/maib/controller/CustomerController.java:10]: I don't care. It is not a part of DB setup, so it have to be excluded

Comment (taragrammatonn) [reply src/main/java/md/maib/controller/CustomerController.java:10]: you can check db layer without rest part

File: src/main/resources/application.yaml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+spring:
+  sql:
+    init:
+      mode: always
```
Comment (Sp1ithy) [inline src/main/resources/application.yaml:4]: Are you sure that is set right **init mode** ? 
Take a look onto Spring Database Initialization

**Comment by Sp1ithy** [comment]
If you have problems with merging branches, take a look onto **trunk based development**.
But firstly be sure that you really understand how all base git operations work properly.

File: pom.xml [hunk 11]
```diff
@@ -19,18 +19,48 @@
 
         <sonar.organization>taragrammatonn</sonar.organization>
         <sonar.host.url>https://sonarcloud.io</sonar.host.url>
+        <testcontainers.version>1.19.3</testcontainers.version>
     </properties>
     <dependencies>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.25.3</version>
```
Comment (taragrammatonn) [inline pom.xml:28]: exclude version to peripheries tag

File: src/main/java/md/maib/entity/Customer.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Column(name = "pan", nullable = false, unique = true)
+    private String email;
+
+
+
+    public Customer() {
+    }
+
+    // Builder pattern
```
Comment (taragrammatonn) [inline src/main/java/md/maib/entity/Customer.java:33]: do not use comments in the code, they often indicate that the code needs to be explained, which means the code is too overengineered - this is not good

Comment (taragrammatonn) [reply src/main/java/md/maib/entity/Customer.java:33]: you can write libraries for public use, then comments in the code are your best friends

File: src/test/java/md/maib/CustomerControllerTest.java (hunk 3 above)
Comment (Sp1ithy) [reply src/test/java/md/maib/CustomerControllerTest.java]: Just an advice, new language feature, but it depends on code style in compony. 
We properly use **var**

File: src/main/java/md/maib/entity/Customer.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+    public String getCvv() {
+        return cvv;
+    }
+
+    public void setCvv(String cvv) {
+        this.cvv = cvv;
+    }
+
+    public String fullName() {
```
Comment (Sp1ithy) [inline src/main/java/md/maib/entity/Customer.java:113]: getFullName()

File: pom.xml [hunk 14]
```diff
@@ -19,18 +19,52 @@
...
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>org.springframework.boot</groupId>
-            <artifactId>spring-boot-starter</artifactId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+        <dependency>
```
Comment (Sp1ithy) [inline pom.xml:34]: How is it used ? You already have postgresql

Comment (jah3) [reply pom.xml:34]: Această dependență facilitează lucrul meu cu clasa CustomerRepository și îmi permite să operez cu metodele CRUD aplicate în CustomerControllerTest.

File: src/test/java/md/maib/CustomerControllerTest.java (hunk 9 above)
Comment (Sp1ithy) [reply src/test/java/md/maib/CustomerControllerTest.java:20]: **CustomerControllerTest** name doesn't match the content

File: pom.xml (hunk 14 above)
Comment (Sp1ithy) [reply pom.xml:34]: Sorry, i've missed the line

File: pom.xml [hunk 15]
```diff
@@ -19,18 +19,52 @@
...
+            <groupId>com.h2database</groupId>
```
Comment (Sp1ithy) [inline pom.xml:35]: How is it used ? You already have postgresql

Comment (jah3) [reply pom.xml:35]: indeed, silly me!

File: src/main/java/md/maib/entity/Customer.java [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    @Column(name = "first_name", nullable = false)
+    private String firstName;
+
+    @Column(name = "last_name", nullable = false)
+    private String lastName;
+
+    @Column(name = "pan", nullable = false, unique = true, length = 16)
+    private String pan;
+
+    @OneToMany(mappedBy = "customer", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
```
Comment (Sp1ithy) [inline src/main/java/md/maib/entity/Customer.java:28]: Read about **FetchType**. 
Eager have to be used very carefully, doesn't see any need for that.

File: src/main/java/md/maib/entity/Customer.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Customer customer = (Customer) o;
+        return Objects.equals(id, customer.id) && Objects.equals(cvv, customer.cvv) && Objects.equals(firstName, customer.firstName) && Objects.equals(lastName, customer.lastName) && Objects.equals(pan, customer.pan) ;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, cvv, firstName, lastName, pan, transactions);
```
Comment (Sp1ithy) [inline src/main/java/md/maib/entity/Customer.java:128]: Transactions - is growing list, making hash gonna grow to be more more and complicate operation, so I not sure that it gonna be used here. @taragrammatonn ?

**Review by taragrammatonn — APPROVED** [review]

**Review by Sp1ithy — APPROVED** [review]

**Pull request merged**